### DISCUSSION TITLE
String generation in  CVC5

### DISCUSSION THREAD

**Discussion by AnkitMU** [cvc5/cvc5#10356] [Q&A]
I'm trying to generate the String value in CVC5. But each time I'm running this SMTlib2 formula in cvc5 I'm getting the same Model Output(please refer below) , although im using distinct function. but its giving me same output and there No diversity in the result, to generate First name im using the constraint as (str.in_re (first_name o) (re.* (re.range "a" "z"))) but it is returning me "aaaaaaaaaaaaaaa". For last name (str.in_re (last_name o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z")))) but it is returning me "AAAAAAAAAAAAAAA".
Am I doing something wrong in the below SMTlib2 formula. How Can I solve this problem. (i) get different result on each model produce (ii) get a more diverse Output as stated in constraint.

(set-logic ALL)
(set-option :produce-models true)
(set-option :strings-exp true)

(declare-fun first_name (Int) String)
(declare-fun last_name (Int) String)
(declare-fun email (Int) String)
(declare-fun address (Int) String)
(declare-fun Account_number (Int) Int)

(define-fun Person ((o Int)) Bool
(and (>= o 1) (<= o 10))
)

(assert
(forall ((o Int)) (=> (Person o)
(and
(> (str.len (first_name o)) 5)
(str.in_re (first_name o) (re.* (re.range "a" "z")))
)
))
)

(assert
(forall ((o Int)) (=> (Person o)
(and
(> (str.len (last_name o)) 5)
(str.in_re (last_name o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))
)
))
)

(assert
(forall ((o Int)) (=> (Person o)
(not (= (first_name o) (last_name o)))
))
)

(assert
(forall ((o Int)) (=> (Person o)
(= (email o) (str.++ (first_name o) "." (last_name o) "@mu.ie"))
))
)

(assert
(forall ((o Int)) (=> (Person o)
(str.suffixof "mu.ie" (email o))
))
)

(assert
(forall ((o Int)) (=> (Person o)
(and
(> (str.len (address o)) 14)
(str.in_re (address o) (re.* (re.union (re.range "A" "Z") (re.range "a" "z"))))
)
))
)

(assert
(forall ((o Int)) (=> (Person o)
(and
(<= (Account_number o) 9999999999)
(>= (Account_number o) 1000000000)
)
))
)

(assert (exists ((o Int)) (Person o)))

(assert (distinct
(str.++ (first_name 1) "1")
(str.++ (first_name 2) "2")
(str.++ (first_name 3) "3")
(str.++ (last_name 1) "1")
(str.++ (last_name 2) "2")
(str.++ (last_name 3) "3")
(str.++ (address 1) "1")
(str.++ (address 2) "2")
(str.++ (address 3) "3")
))
(assert (distinct
(Account_number 1)
(Account_number 2)
(Account_number 3)
))

(check-sat)
(get-value ((first_name 1)(last_name 1)(email 1)(address 1)(Account_number 1)))
(get-value ((first_name 2)(last_name 2)(email 2)(address 2)(Account_number 2)))
(get-value ((first_name 3)(last_name 3)(email 3)(address 3)(Account_number 3)))

OUTPUT
sat
(((first_name 1) "aaaaaaaaaaaaaaa") ((last_name 1) "AAAAAAAAAAAAAAA") ((email 1) "[aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie](mailto:aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie)") ((address 1) "ZAAAAAAAAAAAAAA") ((Account_number 1) 1000000000))
(((first_name 2) "aaaaaaaaaaaaaaa") ((last_name 2) "AAAAAAAAAAAAAAA") ((email 2) "[aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie](mailto:aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie)") ((address 2) "ZAAAAAAAAAAAAAA") ((Account_number 2) 1000000001))
(((first_name 3) "aaaaaaaaaaaaaaa") ((last_name 3) "AAAAAAAAAAAAAAA") ((email 3) "[aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie](mailto:aaaaaaaaaaaaaaa.AAAAAAAAAAAAAAA@mu.ie)") ((address 3) "ZAAAAAAAAAAAAAA") ((Account_number 3) 1000000002))

**Comment by ajreynol** [comment]
What kind of diversity do you mean exactly?  To ensure the first names are distinct, you should use:

`(assert (distinct (first_name 1) (first_name 2) (first_name 3)))`

If you want multiple models, you can try incremental mode and the command `(block-model-values)` or `(block-model)` to block the current model followed by another `(check-sat)`.  See for example https://github.com/cvc5/cvc5/blob/main/test/regress/cli/regress1/model-blocker-values.smt2.

**Reply by AnkitMU** [reply]
Thanks for your reply @ajreynol 
I have tried your suggestion , 

_(check-sat)
(get-value ((first_name 1)(last_name 1)(email 1)(address 1)(Account_number 1)))
(block-model:values)
(get-value ((first_name 2)(last_name 2)(email 2)(address 2)(Account_number 2)))
(get-value ((first_name 3)(last_name 3)(email 3)(address 3)(Account_number 3)))_


(block-model: values) doesn't allow me to check sat or unsat  for _(get-value ((first_name 2)(last_name 2)(email 2)(address 2)(Account_number 2)))_
it gives me an error as 
(error "Cannot get value unless after a SAT or UNKNOWN response.")


I got the final output which is given below:::
sat
(((first_name 1) "aaaaaa") ((last_name 1) "AAAAAA") ((email 1) "aaaaaa.AAAAAA@mu.ie") ((address 1) "AAAAAAAAAAAAAAA") ((Account_number 1) 1000000000))
unsat
(error "Cannot get value unless after a SAT or UNKNOWN response.")
(error "Cannot get value unless after a SAT or UNKNOWN response.")


please help me with your valuable suggestion on this.

**Comment by ajreynol** [comment]
`block-model` is equivalent to an `assert` command.  You need to call `check-sat` again to get the next model.
